spark-plugins/src/test/java/co/cask/hydrator/plugin/spark/test/FeatureGeneratorTest.java

```diff
@@ -0,0 +1,330 @@
+        .set("body", "Use tokenizer to convert strings into tokens").build()
+    );
+
+    // write records to source
+    DataSetManager<Table> inputManager = getDataset(Id.Namespace.DEFAULT, source);
+    MockSource.writeInput(inputManager, input);
+
+    // manually trigger the pipeline
+    WorkflowManager workflowManager = appManager.getWorkflowManager(SmartWorkflow.NAME);
+    workflowManager.start();
+    workflowManager.waitForFinish(5, TimeUnit.MINUTES);
+  }
+
+  private void testSkipGramFeatureGenerator() throws Exception {
+    String source = "skip-gram-feature-generator-source";
+    String sink = "skip-gram-feature-generator-sink";
+    /*
+     * source --> sparkcompute --> sink
+     */
+    ETLBatchConfig etlConfig = ETLBatchConfig.builder("* * * * *")
+      .addStage(new ETLStage("source", MockSource.getPlugin(source, INPUT)))
+      .addStage(new ETLStage("sparkcompute",
+                             new ETLPlugin(SkipGramFeatureGenerator.PLUGIN_NAME, SparkCompute.PLUGIN_TYPE,
+                                           ImmutableMap.of("fileSetName", "training-model",
+                                                           "path", "Model",
+                                                           "outputColumnMapping", "body:result"), null)))
+      .addStage(new ETLStage("sink", MockSink.getPlugin(sink)))
+      .addConnection("source", "sparkcompute")
+      .addConnection("sparkcompute", "sink")
+      .build();
+
+    AppRequest<ETLBatchConfig> appRequest = new AppRequest<>(DATAPIPELINE_ARTIFACT, etlConfig);
+    ApplicationId appId = NamespaceId.DEFAULT.app("SinglePhaseApp");
+    ApplicationManager appManager = deployApplication(appId.toId(), appRequest);
+
+    List<StructuredRecord> input = ImmutableList.of(
+      StructuredRecord.builder(INPUT)
+        .set("offset", 1)
+        .set("body", "Spark").build(),
+      StructuredRecord.builder(INPUT)
+        .set("offset", 2)
+        .set("body", "classes in Java").build(),
+      StructuredRecord.builder(INPUT)
+        .set("offset", 3)
+        .set("body", "Logistic regression to predict data").build()
+    );
+
+    DataSetManager<Table> inputManager = getDataset(Id.Namespace.DEFAULT, source);
+    MockSource.writeInput(inputManager, input);
+
+    // manually trigger the pipeline
+    WorkflowManager workflowManager = appManager.getWorkflowManager(SmartWorkflow.NAME);
+    workflowManager.start();
+    workflowManager.waitForFinish(5, TimeUnit.MINUTES);
+
+    DataSetManager<Table> labeledTexts = getDataset(sink);
+    List<StructuredRecord> structuredRecords = MockSink.readOutput(labeledTexts);
+    Assert.assertEquals(3, structuredRecords.size());
+    List<Double> result;
+    for (StructuredRecord record : structuredRecords) {
+      result = (ArrayList) (record.get("result"));
+      if ((Integer) record.get("offset") == 1) {
+        Assert.assertArrayEquals(new double[]{-0.05517900735139847, -0.13193030655384064, 0.14834064245224},
```

Comment:
0.3 is very large compared to these numbers. I don't think this is actually testing anything worthwhile, especially since the results are based on a 3rd party library that we don't control.

I would remove this and add more tests for the utility code that extracts words from a StructuredRecord. Anything beyond that point is really just testing the spark ml library.